### PR TITLE
GUI/day night cycle

### PR DESCRIPTION
**Pull request by AlemSnyder**
Add day night cycle. It doesn't look nice, but at least it works.

### CODE REVIEW

File: src/gui/scene/helio.cpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    for (int i = 0; i < test_points; i++) {
+        float h = earth_radius + atmosphere_height * i / test_points;
+
+        float c = pow(earth_radius + atmosphere_height, 2) - pow(h, 2);
+        float a = 1 + pow(slope, 2);
+
+        float length =
+            (-2 * h * slope + pow(pow(2 * h * slope, 2) + 4 * a * c, .5)) / (2 * a);
+
+        length =
```
Comment (egelja) [inline src/gui/scene/helio.cpp:89]: use a different var name

**Review by egelja — CHANGES_REQUESTED** [review]

File: resources/shaders/background/Stars.vert [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+main() {
+    age = age_in;
+
+    float cos_angle = -dot(sun_position, pos.xyz)
+        / (length(sun_position) * length(pos.xyz));
+
+    float sun_height = dot(sun_position, vec3(0,0,1))
+        / length(sun_position);
+
+    if (cos_angle + 3*sun_height > 0){
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply resources/shaders/background/Stars.vert:25]: fixed by big refactor

File: resources/shaders/background/Sky.frag [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    vec4 world_space_direction = MVIP * vec4(EyeDirectionScreenSpace, -1, 1);
+
+    //float run = pow(dot(world_space_direction.xy, world_space_direction.xy), 0.5);
+
+    float cos_diffraction_angle = dot(world_space_direction.xyz, sunlight_direction_unit)/ (length(world_space_direction.xyz));
+
+    color = 2 * length(sunlight_color) * color_intensities * (.5 + (pow(cos_diffraction_angle, 2.0) - 0.5) * convolution);
+
+/*
+    float thickness;
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply resources/shaders/background/Sky.frag:43]: fixed by big refactor

File: resources/shaders/background/Sun.vert [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+    vec4 position = star_center_camera_space / abs(star_center_camera_space.w)
+                    + pixel_projection * vec4(star_corner * sun_size, 0, 0);
+    // I have no idea why I have to do this.
+    position.z = star_center_camera_space.z/3;
+
+    gl_Position = position;
+
+/*
+    float height = dot(sun_position, vec3(0,0,1) );
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply resources/shaders/background/Sun.vert:24]: fixed by big refactor

File: src/gui/render/graphics_shaders/sky.cpp [hunk 5]
```diff
@@ -52,72 +73,174 @@ SkyRenderer::render(screen_size_t width, screen_size_t height, GLuint frame_buff
...
+        star_data_.get_num_stars() // number of models
 
     );
 
     glVertexAttribDivisor(0, 0);
     glVertexAttribDivisor(1, 0);
     glDisableVertexAttribArray(0);
     glDisableVertexAttribArray(1);
-    glDisableVertexAttribArray(2);
+    // glDisableVertexAttribArray(2);
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/gui/render/graphics_shaders/sky.cpp:217]: fixed by big refactor

File: src/gui/render/uniform_types.hpp [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+ public:
+    virtual ~LightEnvironment(){};
+
+    virtual glm::vec3 get_light_direction() const = 0;
+    virtual glm::vec3 get_diffuse_light() const = 0;
+    virtual glm::vec3 get_specular_light() const = 0;
+};
+
+class StarRotation {
+    public:
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/gui/render/uniform_types.hpp:20]: fixed by big refactor

**Review by egelja — APPROVED** [review]

**Pull request merged**